pkg/sriov/sriov.go

```diff
@@ -0,0 +1,310 @@
+	}
+
+	if len(conf.ContIFNames) < 1 && len(conf.ContIFNames) != len(conf.HostIFNames) {
+		return fmt.Errorf("number of interface names mismatch ContIFNames: %d HostIFNames: %d", len(conf.ContIFNames), len(conf.HostIFNames))
+	}
+
+	return netns.Do(func(_ ns.NetNS) error {
+		for i, ifName := range conf.ContIFNames {
+			hostIFName := conf.HostIFNames[i]
+
+			// get VF device
+			linkObj, err := s.nLink.LinkByName(ifName)
+			if err != nil {
+				return fmt.Errorf("failed to get netlink device with name %s: %q", ifName, err)
+			}
+
+			// shutdown VF device
+			if err = s.nLink.LinkSetDown(linkObj); err != nil {
+				return fmt.Errorf("failed to set link %s down: %q", ifName, err)
+			}
+
+			// rename VF device
+			err = s.nLink.LinkSetName(linkObj, hostIFName)
+			if err != nil {
+				return fmt.Errorf("failed to rename link %s to host name %s: %q", ifName, hostIFName, err)
+			}
+
+			// move VF device to init netns
+			if err = s.nLink.LinkSetNsFd(linkObj, int(initns.Fd())); err != nil {
+				return fmt.Errorf("failed to move interface %s to init netns: %v", hostIFName, err)
+			}
+		}
+		return nil
+	})
+}
+
+func (s *sriovManager) resetVfVlan(pfName, vfName string) error {
+
+	// get the ifname sriov vf num
+	vfTotal, err := utils.GetSriovNumVfs(pfName)
+	if err != nil {
+		return err
+	}
+
+	if vfTotal <= 0 {
+		return fmt.Errorf("no virtual function in the device: %v", pfName)
+	}
+
+	// Get VF id
+	var vf int
+	idFound := false
+	for vf = 0; vf < vfTotal; vf++ {
+		vfDir := fmt.Sprintf("/sys/class/net/%s/device/virtfn%d/net/%s", pfName, vf, vfName)
+		if _, err := os.Stat(vfDir); !os.IsNotExist(err) {
+			idFound = true
+			break
+		}
+	}
+
+	if !idFound {
+		return fmt.Errorf("failed to get VF id for %s", vfName)
+	}
+
+	pfLink, err := s.nLink.LinkByName(pfName)
+	if err != nil {
+		return fmt.Errorf("master device %s not found", pfName)
+	}
+
+	if err = s.nLink.LinkSetVfVlan(pfLink, vf, 0); err != nil {
+		return fmt.Errorf("failed to reset vlan tag for vf %d: %v", vf, err)
+	}
+	return nil
+}
+
+// ApplyVFConfig configure a VF with parameters given in NetConf
+func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
+
+	pfLink, err := s.nLink.LinkByName(conf.Master)
+	if err != nil {
+		return fmt.Errorf("failed to lookup master %q: %v", conf.Master, err)
+	}
+
+	// 1. Set vlan
+	if conf.Vlan != 0 {
+		if err = s.nLink.LinkSetVfVlan(pfLink, conf.VFID, conf.Vlan); err != nil {
+			return fmt.Errorf("failed to set vf %d vlan: %v", conf.VFID, err)
+		}
+	}
+
+	// 2. Set mac address
+
+	// 3. Set link rate
+
+	return nil
+}
+
+// ResetVFConfig reset a VF with default values
+func (s *sriovManager) ResetVFConfig(conf *sriovtypes.NetConf) error {
+
+	pfLink, err := s.nLink.LinkByName(conf.Master)
+	if err != nil {
+		return fmt.Errorf("failed to lookup master %q: %v", conf.Master, err)
+	}
+
+	// Set vlan to 0
+	if err = s.nLink.LinkSetVfVlan(pfLink, conf.VFID, 0); err != nil {
```

Comment:
I think that this should only be set back to 0 when the VLAN was provided in config, it seems valid to me that someone may configure a VLAN on the VF manually and the plugin should not remove that setting.

Reply:
I think we should cache the old config vlan/link state mode, mac address  before setting a new vf config . So reset config should reset the VF the old config

Reply:
+1 on cache and reset vlan of VF to its original value.